components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1269,6 +1270,224 @@ where
+            .fsm
+            .peer
+            .raft_group
+            .raft
+            .prs()
+            .conf()
+            .voters()
+            .ids()
+            .iter()
+            .filter(|peer_id| {
+                let store_id = region
+                    .get_peers()
+                    .iter()
+                    .find(|p| p.get_id() == *peer_id)
+                    .unwrap()
+                    .get_store_id();
+                !failed_stores.contains(&store_id)
+            })
+            .collect();
+        info!(
+            "enter pre force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+            "alive_voter" => ?expected_alive_voter,
+        );
+
+        if self.fsm.peer.is_leader() {
+            warn!(
+                "pre force leader on leader";
+                "region_id" => self.fsm.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+            );
+            return;
+        }
+
+        // Do not use prevote as prevote won't set `vote` to itself.
+        // When PD issues force leader on two different peer, it may cause
+        // two force leader in same term.
+        self.fsm.peer.raft_group.raft.pre_vote = false;
+        // trigger vote request to all voters, will check the vote result in `check_force_leader`
+        self.fsm.peer.raft_group.campaign().unwrap();
+        assert!(
+            self.fsm.peer.raft_group.raft.state == StateRole::PreCandidate
+                || self.fsm.peer.raft_group.raft.state == StateRole::Candidate
+        );
+        if !self
+            .fsm
+            .peer
+            .raft_group
+            .raft
+            .prs()
+            .votes()
+            .get(&self.fsm.peer.peer_id())
+            .unwrap()
+        {
+            warn!(
+                "pre force leader failed to campaign";
+                "region_id" => self.fsm.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+            );
+            return;
+        }
+
+        self.fsm.peer.force_leader = Some(ForceLeaderState::PreForceLeader {
+            expected_alive_voter,
+        });
+        self.fsm.has_ready = true;
+    }
+
+    fn on_exit_pre_force_leader(&mut self) {
+        info!(
+            "exit pre force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+
+        self.fsm.peer.raft_group.raft.pre_vote = true;
+        self.fsm.peer.force_leader = None;
+    }
+
+    fn on_enter_force_leader(&mut self) {
+        info!(
+            "enter force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+        assert!(!self.fsm.peer.is_leader());
+
+        // become candidate first to increase term
+        self.fsm.peer.raft_group.raft.become_candidate();
+        self.fsm.peer.raft_group.raft.become_leader();
```

Comment:
Instead of calling `become_leader` explicitly, you can cast fake vote to it, and let it become leader internally. So all mechanism like commit by vote, bcastappend right after winning election can be reused.